src/backend/sgx/attestation.rs

```diff
@@ -0,0 +1,346 @@
+            Error::new(
+                ErrorKind::Other,
+                format!("Invalid protobuf request: {:?}. Error: {:?}", req, e),
+            )
+        })?;
+
+        let req_len = (buf_wrtr.len() - size_of::<u32>()) as u32;
+        buf_wrtr[0..size_of::<u32>()].copy_from_slice(&req_len.to_le_bytes());
+
+        // Send Request to AESM daemon
+        stream.write_all(&buf_wrtr)?;
+        stream.flush()?;
+
+        // Receive Response
+        let mut res_len_bytes = [0u8; 4];
+        stream.read_exact(&mut res_len_bytes)?;
+        let res_len = u32::from_le_bytes(res_len_bytes);
+
+        let mut res_bytes = vec![0; res_len as usize];
+        stream.read_exact(&mut res_bytes)?;
+
+        let response = Message::parse_from_bytes(&res_bytes)?;
+
+        Ok(response)
+    }
+}
+
+/// Gets Att Key ID
+fn get_attestation_key_id() -> Result<Vec<u8>, Error> {
+    let mut stream = UnixStream::connect(AESM_SOCKET)?;
+
+    let r = ReqType::AkIdNum;
+    let mut req = Request::new();
+    let mut msg = Request_GetSupportedAttKeyIDNumRequest::new();
+    msg.set_timeout(AESM_REQUEST_TIMEOUT);
+    req.set_getSupportedAttKeyIDNumReq(msg);
+    let pb_msg: Response = r.send_request(req, &mut stream)?;
+
+    let res = pb_msg.get_getSupportedAttKeyIDNumRes();
+    if res.get_errorCode() != 0 {
+        return Err(Error::new(
+            ErrorKind::Other,
+            format!(
+                "Received error code {:?} in GetSupportedAttKeyIDNum",
+                res.get_errorCode()
+            ),
+        ));
+    }
+
+    let num_key_ids = res.get_att_key_id_num();
+    if num_key_ids != 1 {
+        return Err(Error::new(
+            ErrorKind::Other,
+            format!("Unexpected number of key IDs: {} != 1", num_key_ids),
+        ));
+    }
+
+    let expected_buffer_size: u32 = num_key_ids * SGX_KEY_ID_SIZE;
+
+    let mut stream = UnixStream::connect(AESM_SOCKET)?;
+
+    let r = ReqType::AkId;
+    let mut req = Request::new();
+    let mut msg = Request_GetSupportedAttKeyIDsRequest::new();
+    msg.set_timeout(AESM_REQUEST_TIMEOUT);
+    msg.set_buf_size(expected_buffer_size);
+    req.set_getSupportedAttKeyIDsReq(msg);
+
+    let pb_msg: Response = r.send_request(req, &mut stream)?;
+
+    let res = pb_msg.get_getSupportedAttKeyIDsRes();
+
+    if res.get_errorCode() != 0 {
+        return Err(Error::new(
+            ErrorKind::InvalidData,
+            format!("GetSupportedAttKeyIDs: error: {:?}", res.get_errorCode()),
+        ));
+    }
+
+    let key_ids_blob = res.get_att_key_ids();
+    let key_ids: Vec<Vec<u8>> = key_ids_blob
+        .chunks_exact(SGX_KEY_ID_SIZE as usize)
+        .map(Vec::from)
+        .collect();
+
+    if key_ids.len() != 1 {
+        return Err(Error::new(
+            ErrorKind::Other,
+            format!(
+                "GeSupportedAttKeyIDs: invalid count: {} != 1",
+                key_ids.len()
+            ),
+        ));
+    }
+
+    Ok(key_ids.get(0).unwrap().clone())
```

Comment:
Prbly not worth of NAK but could be opened into two separate statements as @rvolosatovs proposed in some other occasion.

Reply:
This is fine, since it's not a fallible operation anymore.